graylog2-server/src/main/java/org/graylog2/filters/ExtractorFilter.java

```diff
@@ -70,34 +83,43 @@ public boolean filter(Message msg) {
         return false;
     }
 
-    private List<Extractor> loadExtractors(final String inputId) {
+    @Subscribe
+    public void handleInputCreate(final InputCreated event) {
+        LOG.debug("Load extractors for input <{}>", event.id());
+        scheduler.schedule(() -> loadExtractors(event.id()), 0, TimeUnit.SECONDS);
+    }
+
+    @Subscribe
+    public void handleInputDelete(final InputDeleted event) {
+        LOG.debug("Removing input from extractors cache <{}>", event.id());
+        extractors.remove(event.id());
+    }
+
+    @Subscribe
+    public void handleInputUpdate(final InputUpdated event) {
+        scheduler.schedule(() -> loadExtractors(event.id()), 0, TimeUnit.SECONDS);
+    }
+
+    private void loadAllExtractors() {
         try {
-            return cache.get(inputId, new Callable<List<Extractor>>() {
-                @Override
-                public List<Extractor> call() throws Exception {
-                    LOG.debug("Re-loading extractors for input <{}> into cache.", inputId);
-
-                    try {
-                        Input input = inputService.find(inputId);
-
-                        List<Extractor> sorted = Lists.newArrayList(inputService.getExtractors(input));
-
-                        Collections.sort(sorted, new Comparator<Extractor>() {
-                            public int compare(Extractor e1, Extractor e2) {
-                                return e1.getOrder().intValue() - e2.getOrder().intValue();
-                            }
-                        });
-
-                        return sorted;
-                    } catch (NotFoundException e) {
-                        LOG.warn("Unable to load input: {}", e.getMessage());
-                        return Collections.emptyList();
-                    }
-                }
-            });
-        } catch (ExecutionException e) {
-            LOG.error("Could not load extractors into cache. Returning empty list.", e);
-            return Collections.emptyList();
+            inputService.all().forEach(input -> loadExtractors(input.getId()));
+        } catch (Exception e) {
+            LOG.error("Unable to load extractors for all inputs", e);
+        }
+    }
+
+    private void loadExtractors(final String inputId) {
```

Comment:
We could also think about `InputServiceImpl#getExtractors()` returning a sorted list in the first place.